bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/scheduler2/InternalSchedulerImpl.java

```diff
@@ -0,0 +1,298 @@
+import java.time.Duration;
+import java.time.Instant;
+import java.time.ZonedDateTime;
+import java.time.temporal.TemporalAdjuster;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.eclipse.smarthome.core.scheduler2.CronJob;
+import org.eclipse.smarthome.core.scheduler2.Scheduler;
+import org.eclipse.smarthome.core.scheduler2.TimeoutException;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * @author Peter Kriens - initial contribution and API
+ * @author Simon Kaufmann - ported to CompletableFuture
+ */
+@Component(name = "org.eclipse.smarthome.scheduler", service = InternalSchedulerImpl.class, immediate = true)
+public class InternalSchedulerImpl implements Scheduler {
+
+    private static final int THREAD_POOL_SIZE = 10;
+    private static final long THREAD_TIMEOUT = 65L;
+
+    private final List<Cron> crons = new ArrayList<>();
+    private final Logger logger = LoggerFactory.getLogger(InternalSchedulerImpl.class);
+
+    private final Clock clock = Clock.systemDefaultZone();
+    private ScheduledExecutorService executor;
+
+    @Activate
+    void activate() {
+        executor = Executors.newScheduledThreadPool(THREAD_POOL_SIZE);
+        ((ThreadPoolExecutor) executor).setKeepAliveTime(THREAD_TIMEOUT, TimeUnit.SECONDS);
+        ((ThreadPoolExecutor) executor).allowCoreThreadTimeOut(true);
+    }
+
+    @Deactivate
+    void deactivate() {
+        List<Runnable> shutdownNow = executor.shutdownNow();
+        if (shutdownNow != null && shutdownNow.size() > 0) {
+            logger.warn("Shutdown executables {}", shutdownNow);
+        }
+    }
+
+    @Override
+    public CompletableFuture<Instant> after(Duration duration) {
+        CompletableFuture<Instant> deferred = new CompletableFuture<>();
+        Instant start = Instant.now();
+        ScheduledFuture<?> schedule = executor.schedule(() -> {
+            deferred.complete(start);
+        }, duration.toMillis(), TimeUnit.MILLISECONDS);
+        deferred.exceptionally((Throwable e) -> {
+            if (e instanceof CancellationException) {
+                schedule.cancel(true);
+            }
+            return null;
```

Comment:
Do we also log the exception somewhere and do we or do we not have any control of whether the schedule should be cancelled when certain types of exceptions occur, see also https://github.com/eclipse/smarthome/issues/3325